File: src/dere_daemon/swarm/git.py
"""Git operations for swarm branch management."""

from __future__ import annotations

import asyncio

from loguru import logger


async def _run_git(working_dir: str, *args: str) -> tuple[int, str, str]:
    """Run git command and return (returncode, stdout, stderr)."""
    proc = await asyncio.create_subprocess_exec(
        "git",
        *args,
        cwd=working_dir,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    stdout, stderr = await proc.communicate()
    return proc.returncode, stdout.decode().strip(), stderr.decode().strip()


async def get_current_branch(working_dir: str) -> str:
    """Get current branch name."""
    returncode, stdout, stderr = await _run_git(
        working_dir, "rev-parse", "--abbrev-ref", "HEAD"
    )
    if returncode != 0:
        raise RuntimeError(f"Failed to get current branch: {stderr}")
    return stdout


async def branch_exists(working_dir: str, branch_name: str) -> bool:
    """Check if a branch exists."""
    returncode, _, _ = await _run_git(
        working_dir, "show-ref", "--verify", "--quiet", f"refs/heads/{branch_name}"
    )
    return returncode == 0


async def create_branch(
    working_dir: str,
    branch_name: str,
    base: str = "HEAD",
) -> str:
    """Create a new branch from base (does not checkout).

    Returns the created branch name.
    """
    # Check if branch already exists
    if await branch_exists(working_dir, branch_name):
        raise RuntimeError(f"Branch '{branch_name}' already exists")

    returncode, _, stderr = await _run_git(
        working_dir, "branch", branch_name, base
    )
    if returncode != 0:
        raise RuntimeError(f"Failed to create branch '{branch_name}': {stderr}")

    logger.info("Created branch '{}' from '{}'", branch_name, base)
    return branch_name


async def checkout_branch(working_dir: str, branch_name: str) -> None:
    """Checkout an existing branch."""
    returncode, _, stderr = await _run_git(working_dir, "checkout", branch_name)
    if returncode != 0:
        raise RuntimeError(f"Failed to checkout branch '{branch_name}': {stderr}")


async def create_and_checkout_branch(
    working_dir: str,
    branch_name: str,
    base: str = "HEAD",
) -> str:
    """Create and checkout a new branch."""
    returncode, _, stderr = await _run_git(
        working_dir, "checkout", "-b", branch_name, base
    )
    if returncode != 0:
        raise RuntimeError(f"Failed to create and checkout branch '{branch_name}': {stderr}")

    logger.info("Created and checked out branch '{}' from '{}'", branch_name, base)
    return branch_name


async def merge_branch(
    working_dir: str,
    source_branch: str,
    target_branch: str,
    no_ff: bool = True,
    message: str | None = None,
) -> tuple[bool, str | None]:
    """Merge source branch into target branch.

    Returns (success, error_message).
    """
    # First checkout target branch
    await checkout_branch(working_dir, target_branch)

    # Build merge command
    args = ["merge", source_branch]
    if no_ff:
        args.append("--no-ff")
    if message:
        args.extend(["-m", message])

    returncode, stdout, stderr = await _run_git(working_dir, *args)

    if returncode != 0:
        # Check if it's a merge conflict
        if "CONFLICT" in stdout or "CONFLICT" in stderr:
            # Abort the merge
            await _run_git(working_dir, "merge", "--abort")
            return False, f"Merge conflict: {stderr or stdout}"
        return False, stderr or stdout

    logger.info("Merged '{}' into '{}'", source_branch, target_branch)
    return True, None


async def delete_branch(working_dir: str, branch_name: str, force: bool = False) -> None:
    """Delete a branch."""
    flag = "-D" if force else "-d"
    returncode, _, stderr = await _run_git(working_dir, "branch", flag, branch_name)
    if returncode != 0:
        raise RuntimeError(f"Failed to delete branch '{branch_name}': {stderr}")


async def get_branch_diff_stat(
    working_dir: str,
    branch: str,
    base: str = "main",
) -> str:
    """Get diff stat between branch and base."""
    returncode, stdout, stderr = await _run_git(
        working_dir, "diff", "--stat", f"{base}...{branch}"
    )
    if returncode != 0:
        raise RuntimeError(f"Failed to get diff stat: {stderr}")
    return stdout


async def has_uncommitted_changes(working_dir: str) -> bool:
    """Check if working directory has uncommitted changes."""
    returncode, stdout, _ = await _run_git(working_dir, "status", "--porcelain")
    return returncode == 0 and bool(stdout.strip())


async def stash_changes(working_dir: str, message: str | None = None) -> bool:
    """Stash any uncommitted changes. Returns True if something was stashed."""
    if not await has_uncommitted_changes(working_dir):
        return False

    args = ["stash", "push"]
    if message:
        args.extend(["-m", message])

    returncode, _, stderr = await _run_git(working_dir, *args)
    if returncode != 0:
        raise RuntimeError(f"Failed to stash changes: {stderr}")

    return True


async def stash_pop(working_dir: str) -> None:
    """Pop the most recent stash."""
    returncode, _, stderr = await _run_git(working_dir, "stash", "pop")
    if returncode != 0:
        raise RuntimeError(f"Failed to pop stash: {stderr}")

File: packages/daemon/src/swarm/git.ts
// Swarm git operations

import { execFile } from "node:child_process";
import { promisify } from "node:util";
import { readdir, readFile, access } from "node:fs/promises";
import { join } from "node:path";
import { constants } from "node:fs";

const execFileAsync = promisify(execFile);

/** Maximum time to wait for a git command in milliseconds */
const GIT_TIMEOUT_MS = 30_000;

export type GitCommandResult = {
  code: number;
  stdout: string;
  stderr: string;
};

export class GitError extends Error {
  readonly code: number;
  readonly stderr: string;
  readonly command: string[];
  readonly workingDir: string;

  constructor(
    message: string,
    options: { code: number; stderr: string; command: string[]; workingDir: string },
  ) {
    super(message);
    this.name = "GitError";
    this.code = options.code;
    this.stderr = options.stderr;
    this.command = options.command;
    this.workingDir = options.workingDir;
  }

  toLogContext(): Record<string, unknown> {
    return {
      error: this.message,
      errorType: this.name,
      code: this.code,
      stderr: this.stderr,
      command: this.command.join(" "),
      workingDir: this.workingDir,
    };
  }
}

/**
 * Validates that the working directory exists and is accessible.
 */
async function validateWorkingDir(workingDir: string): Promise<void> {
  try {
    await access(workingDir, constants.R_OK | constants.X_OK);
  } catch {
    throw new GitError(`Working directory does not exist or is not accessible: ${workingDir}`, {
      code: -1,
      stderr: "ENOENT",
      command: [],
      workingDir,
    });
  }
}

export async function runGitCommand(
  workingDir: string,
  args: string[],
): Promise<GitCommandResult> {
  // Validate inputs
  if (!workingDir || typeof workingDir !== "string") {
    throw new GitError("Working directory is required", {
      code: -1,
      stderr: "Invalid workingDir argument",
      command: args,
      workingDir: workingDir ?? "",
    });
  }

  if (!Array.isArray(args) || args.length === 0) {
    throw new GitError("Git command arguments are required", {
      code: -1,
      stderr: "Invalid args argument",
      command: args ?? [],
      workingDir,
    });
  }

  try {
    const { stdout, stderr } = await execFileAsync("git", args, {
      cwd: workingDir,
      timeout: GIT_TIMEOUT_MS,
      maxBuffer: 10 * 1024 * 1024, // 10MB buffer for large outputs
    });
    return { code: 0, stdout: stdout?.toString() ?? "", stderr: stderr?.toString() ?? "" };
  } catch (error) {
    const err = error as {
      stdout?: string | Buffer;
      stderr?: string | Buffer;
      code?: number;
      killed?: boolean;
      signal?: string;
    };

    // Handle timeout
    if (err.killed && err.signal === "SIGTERM") {
      return {
        code: 124, // Standard timeout exit code
        stdout: err.stdout?.toString() ?? "",
        stderr: `Git command timed out after ${GIT_TIMEOUT_MS}ms: git ${args.join(" ")}`,
      };
    }

    return {
      code: typeof err.code === "number" ? err.code : 1,
      stdout: err.stdout?.toString() ?? "",
      stderr: err.stderr?.toString() ?? String(error),
    };
  }
}

export async function getCurrentBranch(workingDir: string): Promise<string> {
  await validateWorkingDir(workingDir);

  const result = await runGitCommand(workingDir, ["rev-parse", "--abbrev-ref", "HEAD"]);
  if (result.code !== 0) {
    throw new GitError(
      `Failed to get current branch: ${result.stderr || "Unknown error"}`,
      { code: result.code, stderr: result.stderr, command: ["rev-parse", "--abbrev-ref", "HEAD"], workingDir },
    );
  }
  return result.stdout.trim();
}

export async function createBranch(workingDir: string, branchName: string, base: string): Promise<void> {
  await validateWorkingDir(workingDir);

  // Validate branch name (basic sanity check)
  if (!branchName || typeof branchName !== "string" || branchName.includes("..")) {
    throw new GitError(`Invalid branch name: ${branchName}`, {
      code: -1,
      stderr: "Invalid branch name",
      command: ["branch", branchName, base],
      workingDir,
    });
  }

  const exists = await runGitCommand(workingDir, [
    "show-ref",
    "--verify",
    "--quiet",
    `refs/heads/${branchName}`,
  ]);
  if (exists.code === 0) {
    throw new GitError(`Branch '${branchName}' already exists in ${workingDir}`, {
      code: 1,
      stderr: "Branch already exists",
      command: ["branch", branchName, base],
      workingDir,
    });
  }

  const result = await runGitCommand(workingDir, ["branch", branchName, base]);
  if (result.code !== 0) {
    throw new GitError(
      `Failed to create branch '${branchName}' from '${base}': ${result.stderr || "Unknown error"}`,
      { code: result.code, stderr: result.stderr, command: ["branch", branchName, base], workingDir },
    );
  }
}

export type MergeResult = {
  success: boolean;
  error: string | null;
  conflictFiles?: string[];
};

export async function mergeBranch(
  workingDir: string,
  source: string,
  target: string,
  noFf = true,
  message?: string,
): Promise<MergeResult> {
  await validateWorkingDir(workingDir);

  // First, checkout the target branch
  const checkoutResult = await runGitCommand(workingDir, ["checkout", target]);
  if (checkoutResult.code !== 0) {
    return {
      success: false,
      error: `Failed to checkout target branch '${target}': ${checkoutResult.stderr || checkoutResult.stdout}`,
    };
  }

  // Build merge command
  const args = ["merge", source];
  if (noFf) {
    args.push("--no-ff");
  }
  if (message) {
    args.push("-m", message);
  }

  const result = await runGitCommand(workingDir, args);
  if (result.code !== 0) {
    const output = result.stdout + result.stderr;
    const isConflict = output.includes("CONFLICT") || output.includes("Merge conflict");

    if (isConflict) {
      // Try to extract conflict file names
      const conflictMatches = output.match(/CONFLICT.*?: (.+)/g) ?? [];
      const conflictFiles = conflictMatches.map((m) => m.replace(/CONFLICT.*?: /, "").trim());

      // Abort the failed merge to restore clean state
      const abortResult = await runGitCommand(workingDir, ["merge", "--abort"]);
      if (abortResult.code !== 0) {
        return {
          success: false,
          error: `Merge conflict in ${conflictFiles.join(", ") || "files"}, and failed to abort: ${abortResult.stderr}`,
          conflictFiles,
        };
      }

      return {
        success: false,
        error: `Merge conflict between '${source}' and '${target}': ${conflictFiles.join(", ") || output.slice(0, 200)}`,
        conflictFiles,
      };
    }

    return {
      success: false,
      error: `Failed to merge '${source}' into '${target}': ${result.stderr || result.stdout}`,
    };
  }

  return { success: true, error: null };
}

export async function listPlugins(): Promise<Array<{
  name: string;
  version: string;
  description: string;
  has_mcp_servers: boolean;
  mcp_servers: string[];
}>> {
  const pluginsDir = join(process.cwd(), "plugins");
  try {
    const entries = await readdir(pluginsDir, { withFileTypes: true });
    const plugins: Array<{
      name: string;
      version: string;
      description: string;
      has_mcp_servers: boolean;
      mcp_servers: string[];
    }> = [];
    for (const entry of entries) {
      if (!entry.isDirectory() || entry.name.startsWith("_")) {
        continue;
      }
      const pluginJson = join(pluginsDir, entry.name, ".claude-plugin", "plugin.json");
      try {
        const raw = await readFile(pluginJson, "utf-8");
        const data = JSON.parse(raw) as Record<string, unknown>;
        const mcpServers =
          data.mcpServers && typeof data.mcpServers === "object"
            ? Object.keys(data.mcpServers as Record<string, unknown>)
            : [];
        plugins.push({
          name: entry.name,
          version: typeof data.version === "string" ? data.version : "0.0.0",
          description: typeof data.description === "string" ? data.description : "",
          has_mcp_servers: mcpServers.length > 0,
          mcp_servers: mcpServers,
        });
      } catch {
        continue;
      }
    }
    return plugins;
  } catch {
    return [];
  }
}
